### PR TITLE
Add ability to perform a number of attempts on connection before discarding from connection pool

### PR DESCRIPTION
**Pull request by robskillington**
cc @martin-mao @kobolog @xichen2020

### CODE REVIEW

File: client/connection_pool.go [hunk 1]
```diff
@@ -41,6 +41,10 @@ import (
 
 const (
 	channelName = "Client"
+
+	// healthCheckFall is the count of health checks to fail to discard a
+	// connection from the connection pool.
+	healthCheckFall = 2
```
Comment (xichen2020) [inline client/connection_pool.go]: `healthCheckFail`?

File: client/connection_pool.go [hunk 2]
```diff
@@ -210,54 +214,69 @@ func (p *connPool) healthCheckEvery(interval time.Duration, stutter time.Duratio
...
+				var (
+					attempts = healthCheckFall
+					failed   = 0
+					checkErr error
+				)
+				for j := 0; j < attempts; j++ {
+					if err := p.healthCheck(client, p.opts); err != nil {
+						checkErr = err
+						failed++
+						continue
```
Comment (xichen2020) [inline client/connection_pool.go]: Should we add some throttling here perhaps? If the health check failures were due to cpu usage spikes, we'll probably need to wait for the cpu usage to subside before the next health check suceeds

File: client/connection_pool.go (hunk 1 above)
Comment (robskillington) [inline client/connection_pool.go]: Sounds good

File: client/connection_pool.go (hunk 2 above)
Comment (robskillington) [inline client/connection_pool.go]: Sounds good, I'll add a throttle here that is a factor of the host connect timeout.

**Pull request merged**